brainiak/isfc.py

```diff
@@ -1,226 +1,1104 @@
+#  limitations under the License.
+"""Intersubject correlation (ISC) analysis
+
+Functions for computing intersubject correlation (ISC) and related
+analyses (e.g., intersubject funtional correlations; ISFC), as well
+as statistical tests designed specifically for ISC analyses.
+
+"""
+
+# Authors: Sam Nastase, Christopher Baldassano, Mai Nguyen, and Mor Regev
+# Princeton University, 2018
+
+import numpy as np
+from scipy.spatial.distance import squareform
+from scipy.stats import pearsonr, zscore
+from scipy.fftpack import fft, ifft
+import itertools as it
+from brainiak.fcma.util import compute_correlation
+
+def isc(data, pairwise=False, summary_statistic=None, verbose=True):
+    """Intersubject correlation
+
+    For each voxel or ROI, compute the Pearson correlation between each
+    subject's response time series and other subjects' response time series.
+    If pairwise is False (default), use the leave-one-out approach, where
+    correlation is computed between each subject and the average of the other
+    subjects. If pairwise is True, compute correlations between all pairs of
+    subjects. If summary_statistic is None, return N ISC values for N subjects
+    (leave-one-out) or N(N-1)/2 ISC values for each pair of N subjects,
+    corresponding to the upper triangle of the pairwise correlation matrix
+    (see scipy.spatial.distance.squareform). Alternatively, supply either
+    np.mean or np.median to compute summary statistic of ISCs (Fisher Z will
+    be applied and inverted if using mean). Input data should be a list 
+    where each item is a time-points by voxels ndarray for a given subject.
+    Multiple input ndarrays must be the same shape. If a single ndarray is
+    supplied, the last dimension is assumed to correspond to subjects. If 
+    only two subjects are supplied, simply compute Pearson correlation
+    (precludes averaging in leave-one-out approach, and does not apply
+    summary statistic.) Output is an ndarray where the first dimension is
+    the number of subjects or pairs and the second dimension is the number
+    of voxels (or ROIs).
+        
+    The implementation is based on the following publication:
+    
+    .. [Hasson2004] "Intersubject synchronization of cortical activity 
+    during natural vision.", U. Hasson, Y. Nir, I. Levy, G. Fuhrmann,
+    R. Malach, 2004, Science, 303, 1634-1640.
+
+    Parameters
+    ----------
+    data : list or ndarray
+        fMRI data for which to compute ISC
+        
+    pairwise : bool, default: False
+        Whether to use pairwise (True) or leave-one-out (False) approach
+        
+    summary_statistic : None
+        Return all ISCs or collapse using np.mean or np.median
+
+    Returns
+    -------
+    iscs : subjects or pairs by voxels ndarray
+        ISC for each subject or pair (or summary statistic) per voxel
+
+    """
+    
+    # Convert list input to 3d and check shapes
+    if type(data) == list:
```

Comment:
@mihaic @snastase  Do you think it is feasible to have a common function (perhaps implemented in `util`) that validate and standardize input data?

Reply:
@qihongl, we should do that. We already have incipient code in `brainiak.io` and `brainiak.image`.

Reply:
@mihaic Awesome!

Reply:
I can add a function to `brainiak.io` or `brainiak.images` that basically performs the reorganization and stacking operation that's currently inside these functions. As far as I can understand, this should have the flexibility to take in either (a) the iterator from e.g. `load_images`, (b) a simple numpy array, or (c) a list of numpy arrays—and standardize this to `n_samples`(TRs, conditions, etc) by `n_features` (voxels, sensors, etc) by `n_subjects` if necessary. I can figure out how to organize the data intuitively for ISFC-related functions, but if we want this to be a general I/O utility, I may need a bit of help making sure the output organization makes sense for other functionalities. Currently, I'm inferring what the user wants from the input they're providing, and then printing the inferred shape so they can make sure the geometry is as intended. Are folks okay with this approach?

Reply:
This sounds good.

At this moment, we have a single class method that creates the BrainIAK data format from iterable masked images, `MaskedMultiSubjectData.from_masked_images`. We do not use images with 3D voxels (from `load_images`), because masking is always presents in the pipelines used. The BrainIAK data format is a subclass of `ndarray`, so adding option (b) is easy to do. Option (c) is already supported because lists are iterable. You can add support for even more transformations, but I think we should keep the interface precise by adding new class methods as opposed to accepting multiple input types in the same method; e.g., add `MaskedMultiSubjectData.from_useful_input_type`.

Regarding the shape, as discussed before, I am fine with changing the order of dimensions. The only issue I see is inconsistency within BrainIAK, because we have other methods that take data as features by samples (e.g., SRM, HTFA), although they use `ndarray` directly, not `MaskedMultiSubjectData`.

Reply:
For now, I can just change the output shape of MaskedMultiSubjectData.from_masked_images from (n_voxels x n_TRs x n_subjects) to (n_TRs x n_voxels x n_subjects). However, the simplest application of ISC essentially just computes the correlation between two time series; I'd also like the ISC functions to accept, e.g., time series for a single ROI for N subjects. In the context of MaskedMultiSubjectData, this would result in shape (1 x n_TRs x n_subjects); however, I'm not sure this sort of input data really fits the intention of MaskedMultiSubjectData... For example, if I just have a single time series for 10 subjects (e.g., a list of ndarrays), do we expect folks to reorganize everything according to MaskedMultiSubjectData? The only alternatives I can think of are (a) leave some of this functionality in the ISC functions as it's fairly ISC-specific, or (b) write an ISC-specific data (re)organization function that takes care of these cases. Any thoughts? @mihaic @qihongl

Reply:
This ROI data sounds generally useful to me. Unless you are fairly certain other analysis methods would not use it, I think we should define it outside of ISC, e.g., `brainiak.image.ROIData`.